### PR TITLE
test: Add tests case when deltaMergeBySplit + { flushCache | split } occurs simultaneously

### PR DESCRIPTION
**Pull request by breezewish**
### What problem does this PR solve?

Issue Number: ref #5409

Problem Summary: For now we lack some tests to test the behavior when multiple operations occurred at the same time.

### What is changed and how it works?

1. Introduced a new facility called `SyncPoint`, which can be used to control the arbitrary business logical flow to suspend and continue.

2. Added or modified existing ManualCompaction tests based on the new `SyncPoint`:

   - Modified: **BasicManualCompactTest/DuplicatedLogicalId**

     Previously this test uses sleep(500ms) to expect specific code logic is finished. Now this sleep is substituted with SyncPoint, so that expected code logic is always finished and the test can continue without waiting for useless time.

   - Added: **DeltaMergeStoreMergeDeltaBySegmentTest/RetryByFlushCache**

   - Added: **DeltaMergeStoreMergeDeltaBySegmentTest/RetryBySplit**

     The two new cases above tests whether everything works fine when there are mergeDeltaBySplit and flushCache (or split) happening simultaneously, based on the SyncPoint.

3. Added more guards in gtest_main:

   - Failures are now turned into exceptions by default, so that when writing ASSERT in sub-functions it will break the test immediately. This is particularly useful in SyncPoint tests, as when there are failures in previous sync point steps, it is very likely that the rest step will not work as expected, result in infinite wait. By turning into exceptions, we these test cases fail fast.

   - Added stack output when the program is unexpected terminated in fault signals like SIG_ABORT (and also `std::terminate`). This helps locate the code of triggering these faults.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Review by Lloyd-Pottiger — APPROVED** [review]
LGTM

File: dbms/src/TestUtils/TiFlashTestBasic.h [hunk 1]
```diff
@@ -50,23 +51,25 @@ namespace DB
...
+    {                                                                                  \
+        std::string text = fmt::format("Code: {}. {}\n\n", e.code(), e.displayText()); \
+        if (text.find("Stack trace") == std::string::npos)                             \
+            text += fmt::format("Stack trace:\n{}\n", e.getStackTrace().toString());   \
+        FAIL() << text;                                                                \
+    }                                                                                  \
+    catch (...)                                                                        \
+    {                                                                                  \
+        FAIL() << "Uncaught unknown exception";                                        \
     }
```
Comment (JaySon-Huang) [inline dbms/src/TestUtils/TiFlashTestBasic.h:73]: Add a `tryLogCurrentException` in here?

**Comment by breezewish** [comment]
@JaySon-Huang Thanks for the recommendation! I have made updates. Here are the results:

When throwing a `std::exception`:

<pre>
[2022/07/26 14:39:18.803 +08:00] [ERROR] [Exception.cpp:85] ["virtual void DB::DM::tests::DeltaMergeStoreMergeDeltaBySegmentTest_RetryBySplit_Test::TestBody():std::exception. Code: 1001, type: std::exception, e.what() = std::exception"] [thread_id=1]
/Users/breezewish/Work/tiflash/dbms/src/Storages/DeltaMerge/tests/gtest_dm_delta_merge_store.cpp:3884: Failure
Failed
</pre>

When throwing a `DB::Exception`: The throw place is recoverable from the stack

<pre>
/Users/breezewish/Work/tiflash/dbms/src/Storages/DeltaMerge/tests/gtest_dm_delta_merge_store.cpp:3879: Failure
Failed
Code: 0. DB::Exception: foo

Stack trace:

     0x10d2cf7bc	StackTrace::StackTrace() [gtests_dbms+4508202940]
                	dbms/src/Common/StackTrace.cpp:23
     0x10d2cf7f8	StackTrace::StackTrace() [gtests_dbms+4508203000]
                	dbms/src/Common/StackTrace.cpp:22
     0x10077e784	DB::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int) [gtests_dbms+4295010180]
                	dbms/src/Common/Exception.h:41
     0x10077d88c	DB::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int) [gtests_dbms+4295006348]
                	dbms/src/Common/Exception.h:43
     0x1020de8a4	DB::DM::tests::DeltaMergeStoreMergeDeltaBySegmentTest_RetryBySplit_Test::TestBody() [gtests_dbms+4321618084]
<strong>                	dbms/src/Storages/DeltaMerge/tests/gtest_dm_delta_merge_store.cpp:3809</strong>
     0x108c8e300	void testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) [gtests_dbms+4434535168]
                	contrib/googletest/googletest/src/gtest.cc:2443
     0x108c60e34	void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) [gtests_dbms+4434349620]
                	contrib/googletest/googletest/src/gtest.cc:2479
     0x108c60d84	testing::Test::Run() [gtests_dbms+4434349444]
                	contrib/googletest/googletest/src/gtest.cc:2517
     0x108c61d44	testing::TestInfo::Run() [gtests_dbms+4434353476]
                	contrib/googletest/googletest/src/gtest.cc:2693
     0x108c62b50	testing::TestCase::Run() [gtests_dbms+4434357072]
                	contrib/googletest/googletest/src/gtest.cc:2811
     0x108c6cfd0	testing::internal::UnitTestImpl::RunAllTests() [gtests_dbms+4434399184]
                	contrib/googletest/googletest/src/gtest.cc:5177
     0x108c931ac	bool testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) [gtests_dbms+4434555308]
                	contrib/googletest/googletest/src/gtest.cc:2443
     0x108c6cabc	bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) [gtests_dbms+4434397884]
                	contrib/googletest/googletest/src/gtest.cc:2479
     0x108c6c9ac	testing::UnitTest::Run() [gtests_dbms+4434397612]
                	contrib/googletest/googletest/src/gtest.cc:4786
     0x1028d1760	RUN_ALL_TESTS() [gtests_dbms+4329953120]
                	contrib/googletest/googletest/include/gtest/gtest.h:2341
     0x1028d1464	main [gtests_dbms+4329952356]
                	dbms/src/TestUtils/gtests_dbms_main.cpp:77
     0x150d4d088	<unknown symbol>
</pre>

When std::terminate caused by dropping a joinable std::thread:

<pre>
libc++abi: terminating
Received signal Abort trap: 6

     0x111a22b7c	StackTrace::StackTrace() [gtests_dbms+4508232572]
                	dbms/src/Common/StackTrace.cpp:23
     0x111a22bb8	StackTrace::StackTrace() [gtests_dbms+4508232632]
                	dbms/src/Common/StackTrace.cpp:22
     0x10702452c	fault_signal_handler(int) [gtests_dbms+4329981228]
                	dbms/src/TestUtils/gtests_dbms_main.cpp:32
     0x1bf16d4c4	_OSAtomicTestAndClearBarrier [libsystem_platform.dylib+6445864132]
     0x1bf155ee0	__pthread_atfork_prepare_handlers [libsystem_pthread.dylib+6445768416]
     0x1bf090340	<unknown symbol> [libsystem_c.dylib+6444958528]
     0x1bf110b08	<unknown symbol> [libc++abi.dylib+6445484808]
     0x1bf100834	<unknown symbol> [libc++abi.dylib+6445418548]
     0x1bf10fea4	<unknown symbol> [libc++abi.dylib+6445481636]
     0x1bf10fe2c	<unknown symbol> [libc++abi.dylib+6445481516]
     0x10682b444	DB::DM::tests::DeltaMergeStoreMergeDeltaBySegmentTest_RetryBySplit_Test::TestBody() [gtests_dbms+4321621060]
<strong>                	dbms/src/Storages/DeltaMerge/tests/gtest_dm_delta_merge_store.cpp:3810</strong>
     0x10d3e16cc	void testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) [gtests_dbms+4434564812]
                	contrib/googletest/googletest/src/gtest.cc:2443
     0x10d3b4200	void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) [gtests_dbms+4434379264]
                	contrib/googletest/googletest/src/gtest.cc:2479
     0x10d3b4150	testing::Test::Run() [gtests_dbms+4434379088]
                	contrib/googletest/googletest/src/gtest.cc:2517
     0x10d3b5110	testing::TestInfo::Run() [gtests_dbms+4434383120]
                	contrib/googletest/googletest/src/gtest.cc:2693
     0x10d3b5f1c	testing::TestCase::Run() [gtests_dbms+4434386716]
                	contrib/googletest/googletest/src/gtest.cc:2811
     0x10d3c039c	testing::internal::UnitTestImpl::RunAllTests() [gtests_dbms+4434428828]
                	contrib/googletest/googletest/src/gtest.cc:5177
     0x10d3e6578	bool testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) [gtests_dbms+4434584952]
                	contrib/googletest/googletest/src/gtest.cc:2443
     0x10d3bfe88	bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) [gtests_dbms+4434427528]
                	contrib/googletest/googletest/src/gtest.cc:2479
     0x10d3bfd78	testing::UnitTest::Run() [gtests_dbms+4434427256]
                	contrib/googletest/googletest/src/gtest.cc:4786
     0x107024b10	RUN_ALL_TESTS() [gtests_dbms+4329982736]
                	contrib/googletest/googletest/include/gtest/gtest.h:2341
     0x107024814	main [gtests_dbms+4329981972]
                	dbms/src/TestUtils/gtests_dbms_main.cpp:77
     0x155431088	<unknown symbol>
[1]    58158 abort      /Users/breezewish/Work/tiflash/cmake-build-debug/dbms/gtests_dbms
</pre>

File: dbms/src/Common/SyncPoint/SyncChannel.h [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+    }
+
+private:
+    bool has_receiver = false;
+    bool has_data = false;
+    bool is_closing = false;
+
+    std::atomic<int64_t> pending_op = 0;
+
+    std::mutex m_send, m_recv, m_cv;
```
Comment (JaySon-Huang) [inline dbms/src/Common/SyncPoint/SyncChannel.h:104]: I prefer to declare only one var in one line

File: dbms/src/Common/SyncPoint/SyncChannel.h [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+
+#include <Common/SyncPoint/Ctl.h>
+
+#include <boost/noncopyable.hpp>
+#include <mutex>
+
+namespace DB
+{
+
+class SyncPointCtl::SyncChannel : boost::noncopyable
```
Comment (JaySon-Huang) [inline dbms/src/Common/SyncPoint/SyncChannel.h:25]: maybe use the macro `DISALLOW_COPY_AND_MOVE` from `Common/nocopyable.h`

**Review by JaySon-Huang — COMMENTED** [review]
Some small comment

File: dbms/src/Common/SyncPoint/Ctl.cpp [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+void SyncPointCtl::enable(const char * name)
+{
+    {
+        std::unique_lock lock(mu);
+        channels.try_emplace(name,
+                             std::make_pair(
+                                 std::make_shared<SyncChannel>(),
+                                 std::make_shared<SyncChannel>()));
+    }
+    fiu_enable(name, 1, nullptr, 0);
```
Comment (JaySon-Huang) [inline dbms/src/Common/SyncPoint/Ctl.cpp:36]: Does the sync point depends on fiu? I am not sure the reason why we need to call fiu_enable here and fiu_disable in later codes

Comment (breezewish) [reply dbms/src/Common/SyncPoint/Ctl.cpp:36]: Yes, the sync point depends on the fail point to inject at runtime. When fail point is not enabled by conditional compiling instructions, sync point will be also disabled and has no cost.

Comment (JaySon-Huang) [reply dbms/src/Common/SyncPoint/Ctl.cpp:36]: OK, I get it.

**Review by JaySon-Huang — APPROVED** [review]
LGTM with previous minor comments

**Pull request merged**